llvm/lib/Transforms/Vectorize/SandboxVectorizer/DependencyGraph.cpp

```diff
@@ -50,29 +51,137 @@ MemDGNodeIntervalBuilder::make(const Interval<Instruction> &Instrs,
                              cast<MemDGNode>(DAG.getNode(MemBotI)));
 }
 
+DependencyGraph::DependencyType
+DependencyGraph::getRoughDepType(Instruction *FromI, Instruction *ToI) {
+  if (FromI->mayWriteToMemory()) {
+    if (ToI->mayReadFromMemory())
+      return DependencyType::RAW;
+    if (ToI->mayWriteToMemory())
+      return DependencyType::WAW;
+  }
+  if (FromI->mayReadFromMemory()) {
+    if (ToI->mayWriteToMemory())
+      return DependencyType::WAR;
+    if (ToI->mayReadFromMemory())
+      return DependencyType::RAR;
+  }
```

Comment:
Why not else-if in the second outer if for FromI?

Why is ToI checked first for Read in the first-if and for write in the second, any implicit ordering?  If ToI is not mem read/write this code will still go through all the checks?  Can it both read and write memory, if so can we make it simpler, something along these lines where write takes precedence?

```
enum  AccessType {
   NONE,
   READ,
   WRITE,
}  FromMemType = NONE;

// If it could both read and write, write takes precedence
if (FromI->mayWriteToMemory()) 
  FromMemType = AccessType::WRITE;
else if  (FromI->mayReadFromMemory())
  FromMemType = AccessType::READ;

if (FromMemType  != NONE) {
   // Check Write First
    if (ToI->mayWriteToMemory())
       return FromMemType == AccessType::READ ? WAR : WAW;
    if (ToI->mayReadFromMemory())
       return FromMemType == AccessType::READ ? RAR : RAW;  
}

Reply:
> Why not else-if in the second outer if for FromI?

else-if should work. DONE.

> Why is ToI checked first for Read in the first-if and for write in the second, any implicit ordering?

The order matters in the second outer `if` because one of them is a RAR (no dependency) and the other one is a WAR which is a dependency. The order in the first outer `if` shouldn't matter.

>  If ToI is not mem read/write this code will still go through all the checks? 

Currently yes, we could speed it up perhaps by checking `mayReadOrWriteMemory()` for both and skip some of the code if neither is memory. I will add a TODO for it to do this change after we have enough tests and some benchmarks for the DAG.

>  Can it both read and write memory, 

Yes 

>  if so can we make it simpler, something along these lines where write takes precedence?

Yeah, this should also work, but it seems to be doing the same number of checks while being more verbose (because of the enum declaration). I don't feel too strongly about keeping the original one though.